### DISCUSSION TITLE
How to mutate messages when they come in?

### DISCUSSION THREAD

**Discussion by robrechtme** [GetStream/stream-chat-react-native#1047] [Q&A]
Hi!

You can easily manipulate messages before sending them with the `doSendMessageRequest`, but I would like to do the same thing for incoming messages.

I would like to change the `asset_url` and `image_url` of attachments of all messages. What do you think would be the simplest solution? Is there a way to manipulate message objects before they are stored in the `MessageContext`?

**Comment by vishalnarkhede** [comment]
I wouldn't recommend mutating the messages for this purpose. Instead you should customize the relevant UI components. Please take a look at custom attachments doc for this case - https://getstream.io/chat/docs/sdk/reactnative/guides/custom-attachment/

**Reply by robrechtme** [reply]
That's unfortunate since I want the full functionality of `Gallery` and `ImageGallery` but with different `image_url`'s . I will have to copy the components and change a few lines of code.

**Reply by robrechtme** [reply]
It would be nice if you could pass a custom `Image` component to them, just for rendering the images (like the `imageComponent` prop in `ImageGrid`.

**Reply by madsroskar** [reply]
@robrechtme We have an issue on our backlog to make it easier to customize the gallery, gallery images, and a loading indicator separately to make it easier to do this in the future. With the holidays coming up, I don't think it's likely we'll be able to have that change released before the end of the year, but as it looks right now, we should be able to start work on it early in the new year. :)

(For internal tracking/reference: CRNS-484)
